docs/design/plan-modification.md

```diff
@@ -0,0 +1,298 @@
+
+#### Tradeoffs
+
+With easily mocked arguments, `RequiresReplaceIf` functions are unit testable, and, being a field on the `schema.Attribute` struct, easily discoverable. The provider code is reasonably Go-native, as far as the complexities of the various type systems involved allow.
+
+As indicated above, helper functions can be introduced to reduce verbosity.
+
+Compatibility may be an issue if a common use case emerges for RequiresReplace conditions a little more complex than the signature of `RequiresReplaceIfFunc` allows - for example, requiring reading config values. This could be anticipated by having the signature instead be:
+```go
+type RequiresReplaceIfFunc (context.Context, ModifyPlanRequest) bool
+```
+By doing this, however, we would lose the benefits of the `old, new` parameters in reducing verbosity - the provider code would have to repeat the attribute path in order to retrieve the values from `req.Plan` and `req.State`.
+
+### `attr.TypeWithRequiresReplace`
+
+A `ModifyPlan()` or `RequiresReplaceIf()` function could be added to an extension of the `attr.Type` interface:
+
+```go
+type TypeWithRequiresReplace interface {
+  Type
+  
+  RequiresReplaceIf(context.Context, old, new Value) bool
+```
+
+This would allow bundling reusable RequiresReplace behaviour up with a custom type's validation and other behaviours.
+
+Without knowing how custom types will be used by provider developers, this option seems premature, and makes less sense than bundling validation functions with custom types.
+
+### Composition and other helpers
+
+#### `All()`
+
+Similarly to `helper/schema`, the `All` composition helper runs all `RequiresReplaceFunc`s and returns true only if all funcs return true.
+
+```go
+func All(funcs ...RequiresReplaceFunc) bool {}
+```
+
+#### `Sequence()`
+
+Similarly to `helper/schema`, the `Sequence` composition helper runs all `RequiresReplaceFunc`s in sequence, stopping at the first that returns false. 
+
+```go
+func Sequence(funcs ...RequiresReplaceFunc) bool {}
+```
+
+#### `RequiresReplaceIfSet()`
+
+This helper reduces the verbosity of `RequiresReplaceFunc`s that always return false if either old or new value is null or unknown.
+
+Framework implementation:
+
+```go
+func RequiresReplaceIfSet(f RequiresReplaceFunc) RequiresReplaceFunc {
+  return func (ctx context.Context, old, new attr.Value) bool {
+    if old.IsNull() || old.IsUnknown() || new.IsNull() || new.IsUnknown() {
+      return false
+    }
+    return f(ctx, old, new)
+  }
+}
+```
+
+Note that this would require the addition of `IsNull()` and `IsUnknown()` functions to the `attr.Value` interface, since there is at present no way to determine whether a generic `attr.Value` is null or unknown.
+
+## Recommendations
+
+We recommend implementing `schema.Attribute.RequiresReplace`, `schema.Attribute.RequiresReplaceIf`, and the `ResourceWithModifyPlan` interface. Composition and other helpers can be implemented as required.
```

Comment:
I think `RequiresReplaceIf` is on the wrong side of the "usefulness" vs "limitations" barrier for me. I think RequiresReplace makes sense given its utility, and I think the `PlanModifier` @bflad proposed above makes sense (I've actually had that design on my whiteboard for a few months, though I called it `PlanTransformer`), but I think:

1) `RequiresReplaceIf` can easily be a `PlanModifier` helper.
2) Including `RequiresReplaceIf` is going to lead to people who can't do what they want because of its constraints to request other variants be added.

I think the `PlanModifier` approach makes sense because it can meet the needs of `RequiresReplaceIf` and _also_ replace DiffSuppressFuncs.